jetty-ee10/jetty-ee10-quickstart/src/main/java/org/eclipse/jetty/ee10/quickstart/QuickStartDescriptorProcessor.java

```diff
@@ -185,13 +187,39 @@ public void visitContextParam(WebAppContext context, Descriptor descriptor, XmlP
                 Object o = context.getAttribute(MetaInfConfiguration.METAINF_TLDS);
                 if (o instanceof Collection<?>)
                     tlds.addAll((Collection<?>)o);
-                for (String i : values)
+
+                // TODO: this should probably be somewhere else
+                List<Resource.Mount> mountedResources = new ArrayList<>();
+
+                try
                 {
-                    Resource r = Resource.newResource(normalizer.expand(i));
-                    if (r.exists())
-                        tlds.add(r.getURI().toURL());
-                    else
-                        throw new IllegalArgumentException("TLD not found: " + r);
+                    for (String i : values)
+                    {
+                        Resource r;
+                        String entry = normalizer.expand(i);
+                        if (entry.toLowerCase(Locale.ENGLISH).startsWith("jar:file:"))
+                        {
+                            URI uri = URI.create(entry);
+                            mountedResources.add(Resource.mount(uri));
+                            r = Resource.newResource(uri);
+                        }
+                        else
+                        {
+                            r = Resource.newResource(entry);
+                        }
+
+                        if (r.exists())
+                            tlds.add(r.getURI().toURL());
+                        else
+                            throw new IllegalArgumentException("TLD not found: " + r);
+                    }
+                }
+                finally
+                {
+                    for (Resource.Mount mountedResource : mountedResources)
```

Comment:
The place that opens the mount need to close it, not some other code.
It's a reference counting system.

Right now, the QuickStartDescriptorProcessor doesn't open the mounts for arbitrary references.
And the code presented above seems to work for the jetty-maven-plugin IT tests.

Reply:
But the point is this:  the context attribute `MetaInfConfiguration.METAINF_TLDS` is a list of `Resources` that is passed out of this class - the user of that list of `Resources` has no way of knowing whether the mount is still good or not. Yes there is a reference count, but it is on the `Mount`, not on the `Resource`.  Thus, the list of mounts needs to be kept somewhere and then unmounted only when the context is undeployed, so perhaps in `QuickstartConfiguration.deconfigure()`.

Reply:
This is incorrect as well.

The MetaInfConfiguration.METAINF_TLDs is actually a `Collection<URL>` not Resources.

This code block takes a set of String values, performs normalizer expand, and makes that Collection of URLs.
This code is using Resource to do that ....

``` java
if (resource.exists())
   tlds.add(resource.getURI().toURL());
```

It doesn't hold onto the Resource objects, no need for a `deconfigure()` setup.

Reply:
@joakime ah yes, the tlds are URLs, so that's ok. However the `METAINF_RESOURCES` are converted from Strings in the `quickstart-web.xml` file into Resources and then stored as a `Collection<Resource>` in the `METAINF_RESOURCES` context attribute. Those Strings are going to be `jar:file:/xxxx/!META-INF/blah` type urls, so do they need the new special sauce handling for mounts?